debian/mongostat.1
.\" Man page generated from reStructuredText.
.
.TH "MONGOSTAT" "1" "Jun 21, 2018" "4.0" "mongerdb-manual"
.SH NAME
mongerstat \- MongerDB Use Statistics
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Synopsis\fP
.IP \(bu 2
\fI\%Required Access\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Fields\fP
.IP \(bu 2
\fI\%Use\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.IP "Mac OSX Sierra and Go 1.6 Incompatibility"
.sp
Users running on Mac OSX Sierra require the 3.2.10 or newer version
of  mongerstat\&.
.UNINDENT
.UNINDENT
.SH SYNOPSIS
.sp
The \fI\%mongerstat\fP utility provides a quick overview of the
status of a currently running \fBmongerd\fP
or \fBmongers\fP
instance. \fI\%mongerstat\fP is functionally similar to the
UNIX/Linux file system utility \fBvmstat\fP, but provides data regarding
\fBmongerd\fP and \fBmongers\fP instances.
.sp
Run \fI\%mongerstat\fP from the system command line, not the \fBmonger\fP shell.
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
For more information about monitoring MongerDB, see
/administration/monitoring\&.
.sp
For more background on other MongerDB status outputs see:
.INDENT 0.0
.IP \(bu 2
/reference/command/serverStatus
.IP \(bu 2
/reference/command/replSetGetStatus
.IP \(bu 2
/reference/command/dbStats
.IP \(bu 2
/reference/command/collStats
.UNINDENT
.sp
For an additional utility that provides MongerDB metrics see
mongertop\&.
.UNINDENT
.UNINDENT
.SH REQUIRED ACCESS
.sp
In order to connect to a \fBmongerd\fP that enforces authorization
with the \fB\-\-auth\fP option, specify the
\fI\%\-\-username\fP and \fI\%\-\-password\fP options, and the connecting user must have the
\fBserverStatus\fP privilege action on the cluster resources.
.sp
The built\-in role \fBclusterMonitor\fP provides this privilege as
well as other privileges. To create a role with just the privilege to
run \fI\%mongerstat\fP, see create\-role\-for\-mongerstat\&.
.SH OPTIONS
.INDENT 0.0
.TP
.B mongerstat
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBmongerstat\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmongerstat\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-uri <connectionString>
New in version 3.4.6.

.sp
Specify a resolvable URI
connection string for the \fBmongerd\fP to which to
connect.
.sp
The following is the standard
URI connection scheme:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
mongerdb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For detailed explanations of the components of this string, refer to
the
Connection String URI Format
documentation.
.sp
\fBIMPORTANT:\fP
.INDENT 7.0
.INDENT 3.5
The following \fI\%mongerstat\fP options are incompatible with the
\fB\-\-uri\fP option. Instead, specify these options as part of your
\fB\-\-uri\fP connection string when applicable:
.INDENT 0.0
.IP \(bu 2
\fB\-\-host\fP
.IP \(bu 2
\fB\-\-port\fP
.IP \(bu 2
\fB\-\-db\fP
.IP \(bu 2
\fB\-\-username\fP
.IP \(bu 2
\fB\-\-password\fP (when specifying the password as part of the
URI connection string)
.IP \(bu 2
\fB\-\-authenticationDatabase\fP
.IP \(bu 2
\fB\-\-authenticationMechanism\fP
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>, \-h <hostname><:port>
\fIDefault\fP: localhost:27017
.sp
Specifies a resolvable hostname for the \fBmongerd\fP to which to
connect. By default, the \fBmongerstat\fP attempts to connect to a MongerDB
instance running on the localhost on port number \fB27017\fP\&.
.sp
To connect to a replica set, you can specify the set member or
members to report on, as in the following (see also the
\fB\-\-discover\fP flag):
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-\-host <hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Changed in version 3.0.0: If you use IPv6 and use the \fB<address>:<port>\fP format, you must
enclose the portion of an address and port combination in
brackets (e.g. \fB[<address>]\fP).

.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
\fIDefault\fP: 27017
.sp
Specifies the TCP port on which the MongerDB instance listens for
client connections.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
\fIRemoved in version 3.0.\fP
.sp
Enables IPv6 support and allows \fBmongerstat\fP to connect to the
MongerDB instance using an IPv6 network. Prior to MongerDB 3.0, you
had to specify \fI\%\-\-ipv6\fP to use IPv6. In MongerDB 3.0 and later, IPv6
is always enabled.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBmongerd\fP or \fBmongers\fP that has
TLS/SSL support enabled.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
Starting in version 3.4, if \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP is not
specified and you are not using x.509 authentication, the
system\-wide CA certificate store will be used when connecting to an
TLS/SSL\-enabled server.
.sp
If using x.509 authentication, \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP
must be specified.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
\fBVersion 3.2 and earlier:\fP For TLS/SSL connections (\fB\-\-ssl\fP) to
\fBmongerd\fP and \fBmongers\fP, if the \fBmongerstat\fP runs without the
\fI\%\-\-sslCAFile\fP, \fBmongerstat\fP will not attempt
to validate the server certificates. This creates a vulnerability
to expired \fBmongerd\fP and \fBmongers\fP certificates as
well as to foreign processes posing as valid \fBmongerd\fP or
\fBmongers\fP instances. Ensure that you \fIalways\fP specify the
CA file to validate the server certificates in cases where
intrusion is a possibility.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the TLS/SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\%\-\-ssl\fP option to connect
to a \fBmongerd\fP or \fBmongers\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fI\%\-\-sslPEMKeyFile\fP). Use the \fI\%\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBmongerstat\fP will
redact the password from all logging and reporting output.
.sp
If the private key in the PEM file is encrypted and you do not specify
the \fI\%\-\-sslPEMKeyPassword\fP option, the \fBmongerstat\fP will prompt for a passphrase. See
ssl\-certificate\-password\&.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, MongerDB logs as a
warning the use of the invalid certificate.
.sp
Starting in MongerDB 4.0, if you specify
\fB\-\-sslAllowInvalidCertificates\fP or \fBssl.allowInvalidCertificates:
true\fP when using x.509 authentication, an invalid certificate is
only sufficient to establish a TLS/SSL connection but is
\fIinsufficient\fP for authentication.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
For TLS/SSL connections to \fBmongerd\fP and
\fBmongers\fP, avoid using
\fB\-\-sslAllowInvalidCertificates\fP if possible and only use
\fB\-\-sslAllowInvalidCertificates\fP on systems where intrusion is
not possible.
.sp
If the \fBmonger\fP shell (and other
mongerdb\-tools\-support\-ssl) runs with the
\fB\-\-sslAllowInvalidCertificates\fP option, the
\fBmonger\fP shell (and other
mongerdb\-tools\-support\-ssl) will not attempt to validate
the server certificates. This creates a vulnerability to expired
\fBmongerd\fP and \fBmongers\fP certificates as
well as to foreign processes posing as valid
\fBmongerd\fP or \fBmongers\fP instances.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in TLS/SSL certificates. Allows
\fBmongerstat\fP to connect to MongerDB instances even if the hostname in their
certificates do not match the specified hostname.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBmongerstat\fP to use the FIPS mode of the installed OpenSSL
library. Your system must have a FIPS compliant OpenSSL library to use
the \fI\%\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS\-compatible TLS/SSL is
available only in \fI\%MongerDB Enterprise\fP\&. See
/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a MongerDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a MongerDB database
that uses authentication. Use in conjunction with the \fB\-\-username\fP and
\fB\-\-authenticationDatabase\fP options.
.sp
Changed in version 3.0.0: If you do not specify an argument for \fI\%\-\-password\fP, \fBmongerstat\fP returns
an error.

.sp
Changed in version 3.0.2: If you wish \fBmongerstat\fP to prompt the user
for the password, pass the \fI\%\-\-username\fP option without
\fI\%\-\-password\fP or specify an empty string as the \fI\%\-\-password\fP value,
as in \fB\-\-password ""\fP .

.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationDatabase <dbname>
Specifies the database in which the user is created.
See user\-authentication\-database\&.
.sp
\fI\%\-\-authenticationDatabase\fP is required for \fBmongerd\fP
and \fBmongers\fP instances that use authentication\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationMechanism <name>
\fIDefault\fP: SCRAM\-SHA\-1
.sp

debian/changelog
mongerdb (2.5.4) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/12898

 -- Ernie Hershey <ernie.hershey@mongerdb.com>  Thu, 19 Dec 2013 00:41:46 -0500

mongerdb (2.1.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10894

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jun 2012 16:56:28 -0500

mongerdb (1.9.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Thu, 11 Aug 2011 16:56:28 -0500

mongerdb (1.9.1) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jul 2011 16:56:28 -0500

mongerdb (1.9.0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10232

 -- Richard Kreuter <richard@10gen.com>  Sun, 01 May 2011 16:56:28 -0500

mongerdb (1.8.0-rc0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10245

 -- Richard Kreuter <richard@10gen.com>  Tue, 22 Feb 2011 16:56:28 -0500

mongerdb (1.7.6) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10223

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.5) unstable; urgency=low

  * durability journalingin place
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10214

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.4) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10210

 -- Richard Kreuter <richard@10gen.com>  Mon, 20 Dec 2010 16:56:28 -0500

mongerdb (1.7.3) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10195

 -- Richard Kreuter <richard@10gen.com>  Tue, 16 Nov 2010 16:56:28 -0500

mongerdb (1.7.2) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10194

 -- Richard Kreuter <richard@10gen.com>  Fri, 29 Oct 2010 16:56:28 -0500

mongerdb (1.7.1) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10193

 -- Richard Kreuter <richard@10gen.com>  Tue, 28 Sep 2010 16:56:28 -0500

mongerdb (1.7.0) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 03 Sep 2010 16:56:28 -0500

mongerdb (1.5.8) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Tue, 03 Aug 2010 16:56:28 -0500

mongerdb (1.5.7) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 30 Jul 2010 16:56:28 -0500


mongerdb (1.5.6) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10179

 -- Richard Kreuter <richard@10gen.com>  Sat, 24 Jul 2010 16:56:28 -0500

mongerdb (1.5.5) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 16 Jul 2010 16:56:28 -0500

mongerdb (1.5.4) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 2 Jul 2010 16:56:28 -0500

mongerdb (1.5.3) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Thu, 17 Jun 2010 16:56:28 -0500

mongerdb (1.5.2) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10143

 -- Richard Kreuter <richard@10gen.com>  Wed, 27 May 2010 16:56:28 -0500

mongerdb (1.5.1) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10142

 -- Richard Kreuter <richard@10gen.com>  Wed, 3 May 2010 16:56:28 -0500

mongerdb (1.5.0) unstable; urgency=low

  * replication w & real-time, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10125

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500


mongerdb (1.3.5) unstable; urgency=low

  * bug fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500

mongerdb (1.3.4) unstable; urgency=low

  * bufg fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 17 Mar 2010 16:56:28 -0500

mongerdb (1.3.3) unstable; urgency=low

  * geo

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongerdb (1.3.2) unstable; urgency=low

  * munged debian files

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongerdb (1.3.1) unstable; urgency=low

  * Initial release

 -- Kristina Chodorow <kristina@10gen.com>  Tue, 07 Apr 2009 10:18:58 -0400



src/monger/gotools/src/github.com/mongodb/mongo-tools/mongostat/mongostat.go
// Copyright (C) MongoDB, Inc. 2014-present.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License. You may obtain
// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

// Package mongerstat provides an overview of the status of a currently running mongerd or mongers instance.
package mongerstat

import (
	"fmt"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/mongerdb/monger-tools-common/db"
	"github.com/mongerdb/monger-tools-common/log"
	"github.com/mongerdb/monger-tools-common/options"
	"github.com/mongerdb/monger-tools/mongerstat/stat_consumer"
	"github.com/mongerdb/monger-tools/mongerstat/stat_consumer/line"
	"github.com/mongerdb/monger-tools/mongerstat/status"
	"go.mongerdb.org/monger-driver/bson"
)

// MongerStat is a container for the user-specified options and
// internal cluster state used for running mongerstat.
type MongerStat struct {
	// Generic monger tool options.
	Options *options.ToolOptions

	// Mongerstat-specific output options.
	StatOptions *StatOptions

	// How long to sleep between printing the rows, and polling the server.
	SleepInterval time.Duration

	// New nodes can be "discovered" by any other node by sending a hostname
	// on this channel.
	Discovered chan string

	// A map of hostname -> NodeMonitor for all the hosts that
	// are being monitored.
	Nodes map[string]*NodeMonitor

	// ClusterMonitor to manage collecting and printing the stats from all nodes.
	Cluster ClusterMonitor

	// Mutex to handle safe concurrent adding to or looping over discovered nodes.
	nodesLock sync.RWMutex
}

// ConfigShard holds a mapping for the format of shard hosts as they
// appear in the config.shards collection.
type ConfigShard struct {
	Id   string `bson:"_id"`
	Host string `bson:"host"`
}

// NodeMonitor contains the connection pool for a single host and collects the
// mongerstat data for that host on a regular interval.
type NodeMonitor struct {
	host, alias     string
	sessionProvider *db.SessionProvider

	// The time at which the node monitor last processed an update successfully.
	LastUpdate time.Time

	// The most recent error encountered when collecting stats for this node.
	Err error
}

// SyncClusterMonitor is an implementation of ClusterMonitor that writes output
// synchronized with the timing of when the polling samples are collected.
// Only works with a single host at a time.
type SyncClusterMonitor struct {
	// Channel to listen for incoming stat data
	ReportChan chan *status.ServerStatus

	// Channel to listen for incoming errors
	ErrorChan chan *status.NodeError

	// Creates and consumes StatLines using ServerStatuses
	Consumer *stat_consumer.StatConsumer
}

// ClusterMonitor maintains an internal representation of a cluster's state,
// which can be refreshed with calls to Update(), and dumps output representing
// this internal state on an interval.
type ClusterMonitor interface {
	// Monitor() triggers monitoring and dumping output to begin
	// sleep is the interval to sleep between output dumps.
	// returns an error if it fails, and nil when monitoring ends
	Monitor(sleep time.Duration) error

	// Update signals the ClusterMonitor implementation to refresh its internal
	// state using the data contained in the provided ServerStatus.
	Update(stat *status.ServerStatus, err *status.NodeError)
}

// AsyncClusterMonitor is an implementation of ClusterMonitor that writes output
// gotten from polling samples collected asynchronously from one or more servers.
type AsyncClusterMonitor struct {
	Discover bool

	// Channel to listen for incoming stat data
	ReportChan chan *status.ServerStatus

	// Channel to listen for incoming errors
	ErrorChan chan *status.NodeError

	// Map of hostname -> latest stat data for the host
	LastStatLines map[string]*line.StatLine

	// Mutex to protect access to LastStatLines
	mapLock sync.RWMutex

	// Creates and consumes StatLines using ServerStatuses
	Consumer *stat_consumer.StatConsumer
}

// Update refreshes the internal state of the cluster monitor with the data
// in the StatLine. SyncClusterMonitor's implementation of Update blocks
// until it has written out its state, so that output is always dumped exactly
// once for each poll.
func (cluster *SyncClusterMonitor) Update(stat *status.ServerStatus, err *status.NodeError) {
	if err != nil {
		cluster.ErrorChan <- err
		return
	}
	cluster.ReportChan <- stat
}

// Monitor waits for data on the cluster's report channel. Once new data comes
// in, it formats and then displays it to stdout.
func (cluster *SyncClusterMonitor) Monitor(_ time.Duration) error {
	receivedData := false
	for {
		var statLine *line.StatLine
		var ok bool
		select {
		case stat := <-cluster.ReportChan:
			statLine, ok = cluster.Consumer.Update(stat)
			if !ok {
				continue
			}
		case err := <-cluster.ErrorChan:
			if !receivedData {
				return err
			}
			statLine = &line.StatLine{
				Error:  err,
				Fields: map[string]string{"host": err.Host},
			}
		}
		receivedData = true
		if cluster.Consumer.FormatLines([]*line.StatLine{statLine}) {
			return nil
		}
	}
}

// updateHostInfo updates the internal map with the given StatLine data.
// Safe for concurrent access.
func (cluster *AsyncClusterMonitor) updateHostInfo(stat *line.StatLine) {
	cluster.mapLock.Lock()
	defer cluster.mapLock.Unlock()
	host := stat.Fields["host"]
	cluster.LastStatLines[host] = stat
}

// printSnapshot formats and dumps the current state of all the stats collected.
// returns whether the program should now exit
func (cluster *AsyncClusterMonitor) printSnapshot() bool {
	cluster.mapLock.RLock()
	defer cluster.mapLock.RUnlock()
	lines := make([]*line.StatLine, 0, len(cluster.LastStatLines))
	for _, stat := range cluster.LastStatLines {
		lines = append(lines, stat)
	}
	if len(lines) == 0 {
		return false
	}
	return cluster.Consumer.FormatLines(lines)
}

// Update sends a new StatLine on the cluster's report channel.
func (cluster *AsyncClusterMonitor) Update(stat *status.ServerStatus, err *status.NodeError) {
	if err != nil {
		cluster.ErrorChan <- err
		return
	}
	cluster.ReportChan <- stat
}

// The Async implementation of Monitor starts the goroutines that listen for incoming stat data,
// and dump snapshots at a regular interval.
func (cluster *AsyncClusterMonitor) Monitor(sleep time.Duration) error {
	select {
	case stat := <-cluster.ReportChan:
		cluster.Consumer.Update(stat)
	case err := <-cluster.ErrorChan:
		// error out if the first result is an error
		return err
	}

	go func() {
		for {
			select {
			case stat := <-cluster.ReportChan:
				statLine, ok := cluster.Consumer.Update(stat)
				if ok {
					cluster.updateHostInfo(statLine)
				}
			case err := <-cluster.ErrorChan:
				cluster.updateHostInfo(&line.StatLine{
					Error:  err,
					Fields: map[string]string{"host": err.Host},
				})
			}
		}
	}()

	for range time.Tick(sleep) {
		if cluster.printSnapshot() {
			break
		}
	}
	return nil
}

// NewNodeMonitor copies the same connection settings from an instance of
// ToolOptions, but monitors fullHost.
func NewNodeMonitor(opts options.ToolOptions, fullHost string) (*NodeMonitor, error) {
	optsCopy := opts
	host, port := parseHostPort(fullHost)
	optsCopy.Connection.Host = host
	optsCopy.Connection.Port = port
	uriCopy := *opts.URI
	newCS, err := rewriteURI(uriCopy.ConnectionString, fullHost)
	if err != nil {
		return nil, err
	}
	uriCopy.ConnectionString = newCS
	optsCopy.URI = &uriCopy
	optsCopy.Direct = true
	sessionProvider, err := db.NewSessionProvider(optsCopy)
	if err != nil {
		return nil, err
	}
	return &NodeMonitor{
		host:            fullHost,
		sessionProvider: sessionProvider,
		LastUpdate:      time.Now(),
		Err:             nil,
	}, nil
}

func rewriteURI(oldURI, newAddress string) (string, error) {
	u, err := url.Parse(oldURI)
	if err != nil {
		return "", err
	}
	u.Host = newAddress
	return u.String(), nil
}

func (node *NodeMonitor) Disconnect() {
	node.sessionProvider.Close()
}

// Report collects the stat info for a single node and sends found hostnames on
// the "discover" channel if checkShards is true.
func (node *NodeMonitor) Poll(discover chan string, checkShards bool) (*status.ServerStatus, error) {
	stat := &status.ServerStatus{}
	log.Logvf(log.DebugHigh, "getting session on server: %v", node.host)
	session, err := node.sessionProvider.GetSession()
	if err != nil {
		log.Logvf(log.DebugLow, "got error getting session to server %v", node.host)
		return nil, err
	}
	log.Logvf(log.DebugHigh, "got session on server: %v", node.host)

	result := session.Database("admin").RunCommand(nil, bson.D{{"serverStatus", 1}, {"recordStats", 0}})
	err = result.Err()
	if err != nil {
		log.Logvf(log.DebugLow, "got error calling serverStatus against server %v", node.host)
		return nil, err
	}
	tempBson, err := result.DecodeBytes()
	if err != nil {
		log.Logvf(log.Always, "Encountered error decoding serverStatus: %v\n", err)
		return nil, fmt.Errorf("Error decoding serverStatus: %v\n", err)
	}
	err = bson.Unmarshal(tempBson, &stat)
	if err != nil {
		log.Logvf(log.Always, "Encountered error reading serverStatus: %v\n", err)
		return nil, fmt.Errorf("Error reading serverStatus: %v\n", err)
	}
	// The flattened version is required by some lookup functions
	statMap := make(map[string]interface{})
	err = result.Decode(&statMap)
	if err != nil {
		return nil, fmt.Errorf("Error flattening serverStatus: %v\n", err)
	}
	stat.Flattened = status.Flatten(statMap)

	node.Err = nil
	stat.SampleTime = time.Now()

	if stat.Repl != nil && discover != nil {
		for _, host := range stat.Repl.Hosts {
			discover <- host
		}
		for _, host := range stat.Repl.Passives {
			discover <- host
		}
	}
	node.alias = stat.Host
	stat.Host = node.host
	if discover != nil && stat != nil && status.IsMongers(stat) && checkShards {
		log.Logvf(log.DebugLow, "checking config database to discover shards")
		shardCursor, err := session.Database("config").Collection("shards").Find(nil, bson.M{}, nil)
		if err != nil {
			return nil, fmt.Errorf("error discovering shards: %v", err)
		}
		shard := ConfigShard{}
		for shardCursor.Next(nil) {
			if cursorErr := shardCursor.Decode(&shard); cursorErr != nil {
				return nil, fmt.Errorf("error decoding shard info: %v", err)
			}
			shardHosts := strings.Split(shard.Host, ",")
			for _, shardHost := range shardHosts {
				discover <- shardHost
			}
		}
		shardCursor.Close(nil)
	}

	return stat, nil
}

// Watch continuously collects and processes stats for a single node on a
// regular interval. At each interval, it triggers the node's Poll function
// with the 'discover' channel.
func (node *NodeMonitor) Watch(sleep time.Duration, discover chan string, cluster ClusterMonitor) {
	var cycle uint64
	for ticker := time.Tick(sleep); ; <-ticker {
		log.Logvf(log.DebugHigh, "polling server: %v", node.host)
		stat, err := node.Poll(discover, cycle%10 == 0)

		if stat != nil {
			log.Logvf(log.DebugHigh, "successfully got statline from host: %v", node.host)
		}
		var nodeError *status.NodeError
		if err != nil {
			nodeError = status.NewNodeError(node.host, err)
		}
		cluster.Update(stat, nodeError)
		cycle++
	}
}

func parseHostPort(fullHostName string) (string, string) {
	if colon := strings.LastIndex(fullHostName, ":"); colon >= 0 {
		return fullHostName[0:colon], fullHostName[colon+1:]
	}
	return fullHostName, "27017"
}

// AddNewNode adds a new host name to be monitored and spawns the necessary
// goroutine to collect data from it.
func (mstat *MongerStat) AddNewNode(fullhost string) error {
	mstat.nodesLock.Lock()
	defer mstat.nodesLock.Unlock()

	// Remove the 'shardXX/' prefix from the hostname, if applicable
	pieces := strings.Split(fullhost, "/")
	fullhost = pieces[len(pieces)-1]

	if _, hasKey := mstat.Nodes[fullhost]; hasKey {
		return nil
	}
	for _, node := range mstat.Nodes {
		if node.alias == fullhost {
			return nil
		}
	}
	log.Logvf(log.DebugLow, "adding new host to monitoring: %v", fullhost)
	// Create a new node monitor for this host
	node, err := NewNodeMonitor(*mstat.Options, fullhost)
	if err != nil {
		return err
	}
	mstat.Nodes[fullhost] = node
	go node.Watch(mstat.SleepInterval, mstat.Discovered, mstat.Cluster)
	return nil
}

// Run is the top-level function that starts the monitoring
// and discovery goroutines
func (mstat *MongerStat) Run() error {
	if mstat.Discovered != nil {
		go func() {
			for {
				newHost := <-mstat.Discovered
				err := mstat.AddNewNode(newHost)
				if err != nil {
					log.Logvf(log.Always, "can't add discovered node %v: %v", newHost, err)
				}
			}
		}()
	}
	return mstat.Cluster.Monitor(mstat.SleepInterval)
}
